baybe/surrogates/bandit.py

```diff
@@ -0,0 +1,153 @@
+from baybe.surrogates.base import Surrogate
+from baybe.targets.binary import _NEGATIVE_VALUE_COMP, _POSITIVE_VALUE_COMP
+from baybe.utils.random import temporary_seed
+
+if TYPE_CHECKING:
+    from botorch.models.model import Model
+    from botorch.posteriors import TorchPosterior
+    from torch import Tensor
+
+
+@define
+class BernoulliMultiArmedBanditSurrogate(Surrogate):
+    """A multi-armed bandit model with Bernoulli likelihood and beta prior."""
+
+    joint_posterior: ClassVar[bool] = False
+    # See base class.
+
+    supports_transfer_learning: ClassVar[bool] = False
+    # See base class.
+
+    prior: BetaPrior = field(factory=lambda: BetaPrior(1, 1))
+    """The beta prior for the win rates of the bandit arms. Uniform by default."""
+
+    # TODO: type should be `torch.Tensor | None` but is currently
+    #   omitted due to: https://github.com/python-attrs/cattrs/issues/531
+    _win_lose_counts = field(init=False, default=None, eq=False)
+    """Sufficient statistics for the Bernoulli likelihood model."""
+
+    def posterior_modes(self) -> Tensor:
+        """Compute the posterior mode win rates for all arms.
+
+        Returns:
+            A tensor of length ``N`` containing the posterior mode estimates of the win
+            rates, where ``N`` is the number of bandit arms.
+            Contains ``float('nan')`` for arms with undefined mode.
+        """
+        from torch.distributions import Beta
+
+        return Beta(*self._posterior_beta_parameters()).mode
+
+    def posterior_means(self) -> Tensor:
+        """Compute the posterior mean win rates for all arms.
+
+        Returns:
+            A tensor of length ``N`` containing the posterior mean estimates of the win
+            rates, where ``N`` is the number of bandit arms.
+        """
+        from torch.distributions import Beta
+
+        return Beta(*self._posterior_beta_parameters()).mean
+
+    def _posterior_beta_parameters(self) -> Tensor:
+        """Compute the posterior parameters of the beta distribution.
+
+        Raises:
+            ModelNotTrainedError: If accessed before the model was trained.
+
+        Returns:
+            A tensors of shape ``(2, N)`` containing the posterior beta parameters,
+            where ``N`` is the number of bandit arms.
+        """
+        if self._win_lose_counts is None:
+            raise ModelNotTrainedError(
+                f"'{self.__class__.__name__}' must be trained before posterior "
+                f"information can be accessed."
+            )
+        return self._win_lose_counts + self.prior.to_tensor().unsqueeze(-1)
+
+    def to_botorch(self) -> Model:  # noqa: D102
+        # See base class.
+
+        # We register the sampler on the fly to avoid eager loading of torch
+
+        from botorch.sampling.base import MCSampler
+        from botorch.sampling.get_sampler import GetSampler
+        from torch.distributions import Beta
+
+        class CustomMCSampler(MCSampler):
```

Comment:
What is this class doing here resp. how does the to_botorch function work in this scenario?

Reply:
This has nothing to do with `to_botorch` actually! The only reason why it's in that function is because we need to lazily regist the samples, and since `to_botorch` is the natural transition point where botorch is enountered for the first time from the perspective of the surrogate, we do the registration there. But for the understanding: pretend it wasn't there. Will resolve but reopen if necessary